flink-streaming-java/src/main/java/org/apache/flink/streaming/api/windowing/triggers/ProcessingTimeoutTrigger.java

```diff
@@ -0,0 +1,100 @@
+package org.apache.flink.streaming.api.windowing.triggers;
+
+import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.api.common.state.ValueState;
+import org.apache.flink.api.common.state.ValueStateDescriptor;
+import org.apache.flink.api.common.typeutils.base.LongSerializer;
+import org.apache.flink.streaming.api.windowing.windows.Window;
+
+/**
+ * A {@link Trigger} that can turn any {@link Trigger} into a timeout {@code Trigger}.
+ *
+ * <p>Each record arriving will emit a ProcessingTimeTimer withing the interval,
+ * you can control if the timer will be registered periodic for each event arriving,
+ * by the continualInterval flag.
+ * you can control if the state will be cleared after reach the timeout, by the
+ * shouldClearAtTimeout flag.
+ *
+ * @param <T> The type of elements on which this trigger can operate.
+ * @param <W> The type of {@link Window Windows} on which this trigger can operate.
+ */
+
+@PublicEvolving
+public class ProcessingTimeoutTrigger<T, W extends Window> extends Trigger<T, W> {
+
+	private static final long serialVersionUID = 1L;
+
+	private Trigger<T, W> nestedTrigger;
+	private final long interval;
+	private final boolean continualInterval;
+	private final boolean shouldClearAtTimeout;
+
+	private final ValueStateDescriptor<Long> timeoutStateDesc;
+
+	private ProcessingTimeoutTrigger(Trigger<T, W> nestedTrigger, long interval, boolean continualInterval, boolean shouldClearAtTimeout) {
+		this.nestedTrigger = nestedTrigger;
+		this.interval = interval;
+		this.continualInterval = continualInterval;
+		this.shouldClearAtTimeout = shouldClearAtTimeout;
+		this.timeoutStateDesc = new ValueStateDescriptor<Long>("timeout", LongSerializer.INSTANCE);
+	}
+
+	@Override
+	public TriggerResult onElement(T element, long timestamp, W window, TriggerContext ctx) throws Exception {
+		TriggerResult triggerResult = this.nestedTrigger.onElement(element, timestamp, window, ctx);
+		if (triggerResult.isFire()) {
+			this.clear(window, ctx);
+			return triggerResult;
+		}
+		
+		ValueState<Long> timeoutState = ctx.getPartitionedState(this.timeoutStateDesc);
+		long nextFireTimestamp = ctx.getCurrentProcessingTime() + this.interval;
+
+		if (timeoutState.value() != null && continualInterval) {
+			ctx.deleteProcessingTimeTimer(timeoutState.value());
+			timeoutState.clear();
+		}
+
+		if (timeoutState.value() == null) {
```

Comment:
`timeoutState.value()` should only be called once in this method. Every access is potentially costly because we might have to go out to RocksDB for state access if the RocksDB state backend is used.

Reply:
You absolutely right, i changed it.